node/src/chain/acceptor.rs

```diff
@@ -171,25 +176,85 @@ impl<DB: database::DB, VM: vm::VMExecution, N: Network> Acceptor<N, DB, VM> {
+        // FIXME: Change `changed_provisioners` to return the status of changes.
+        // That way, we can handle everything here without asking for
+        // `get_provisioner`
+        let changed_prov = Self::changed_provisioners(blk, txs)?;
+        if changed_prov.is_empty() {
+            provisioners_list.remove_previous();
+        } else {
+            let mut new_prov = provisioners_list.current().clone();
+            for pk in changed_prov {
+                match vm.get_provisioner(pk.inner())? {
+                    Some(stake) if stake.value() >= MINIMUM_STAKE => {
+                        debug!("SELECTIVE: New stake {stake:?}");
+                        let replaced = new_prov
+                            .replace_stake(pk, stake)
+                            .ok_or(anyhow::anyhow!(
+                                "Replaced a not existing stake"
+                            ))?;
+                        debug!("SELECTIVE: Old stake {replaced:?}");
+                    }
+                    _ => {
+                        let removed = new_prov.remove_stake(&pk).ok_or(
+                            anyhow::anyhow!("Removed a not existing stake"),
+                        )?;
+                        debug!("SELECTIVE: Removed stake {removed:?}");
+                    }
+                }
+            }
+            // Update new prov
+            provisioners_list.update_and_swap(new_prov);
+        }
+        Ok(())
+    }
+
+    fn changed_provisioners(
+        blk: &Block,
+        txs: &[SpentTransaction],
+    ) -> Result<Vec<PublicKey>> {
         // Update provisioners at every epoch (where new stakes take effect)
         if blk.header().height % EPOCH == 0 {
-            return true;
+            anyhow::bail!("Epoch changed");
         }
+
+        let mut changed_provisioners = vec![];
+
         // Update provisioners if a slash has been applied
-        if blk
-            .header()
-            .failed_iterations
-            .to_missed_generators_bytes()
-            .next()
-            .is_some()
+        for bytes in blk.header().failed_iterations.to_missed_generators_bytes()
         {
-            return true;
-        };
-        // Update provisioners if there is a processed unstake transaction
-        txs.iter().filter(|t| t.err.is_none()).any(|t| {
-            matches!(&t.inner.inner.call, Some((STAKE_CONTRACT, f, _)) if f == "unstake")
-        })
+            changed_provisioners.push(bytes.0.try_into().map_err(|e| {
+                anyhow::anyhow!("Cannot deserialize bytes {e:?}")
+            })?);
+        }
+
+        // FIX_ME: This relies on stakeContract to being called only by the
+        // transfer contract
+        let unstake_calldata = txs
+            .iter()
+            .filter(|t| t.err.is_none())
+            .filter_map(|t| match &t.inner.inner.call {
+                Some((STAKE_CONTRACT, f, data)) if f == "unstake" => Some(data),
+                _ => None,
+            });
```

Comment:
You're calling `filter_map` directly after `filter`. I think this can probably be done in the same `filter_map` call.

Reply:
I did it on purpose to remark that we filter out any failed tx